tcms/core/tests/test_ajax.py

```diff
@@ -229,3 +228,48 @@ def _assert_tags(self, response):
         self.assertContains(response, '>4</a>')
         self.assertContains(response, '>5</a>')
         self.assertContains(response, '>6</a>')
+
+
+class Test_Tag_Counter(test.TestCase):
+
+    @classmethod
+    def setUpTestData(cls):
+        cls.tag_one = TagFactory()
+        cls.tag_two = TagFactory()
+
+    def test_tag_count(self):
+        test_case_one = TestCaseFactory()
+        test_case_two = TestCaseFactory()
+        test_case_three = TestCaseFactory()
+
+        # adding tag_one to all 3 test cases
+        test_case_one.add_tag(self.tag_one)
+        test_case_two.add_tag(self.tag_one)
+        test_case_three.add_tag(self.tag_one)
+
+        # adding tag_two to only 2 of the test cases
+        test_case_one.add_tag(self.tag_two)
+        test_case_two.add_tag(self.tag_two)
+
+        test_case_tags = TestCaseTag.objects.filter(
+            tag__in=[self.tag_one, self.tag_two]).values('tag').annotate(
+            num_cases=Count('tag')).order_by('tag')
+
+        case_tag_counter = _TagCounter('num_cases', test_case_tags)
+        count_for_tag_one = case_tag_counter.calculate_tag_count(self.tag_one)
+        count_for_tag_two = case_tag_counter.calculate_tag_count(self.tag_two)
+
+        self.assertEqual(count_for_tag_one, 3)
+        self.assertEqual(count_for_tag_two, 2)
+
+    def test_tag_count_empty_query(self):
+        test_case_tags = TestCaseTag.objects.filter(
+            tag__in=[self.tag_one, self.tag_two]).values('tag').annotate(
```

Comment:
In this case you may use `tag=-1` or even `objects.none()` to save on some typing and multiline indentations.

If the query is empty it doesn't matter how much annotations you make, it is still going to be empty.